pyiron_base/generic/hdfstub.py

```diff
@@ -0,0 +1,101 @@
+"""
+Convenience class to lazily read values from HDF.
+"""
+
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+__author__ = "Marvin Poul"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "1.0"
+__maintainer__ = "Marvin Poul"
+__email__ = "poul@mpie.de"
+__status__ = "production"
+__date__ = "Apr 26, 2021"
+
+
+class HDFStub:
+    """
+    Provides lazy loading of data from HDF.
+
+    Instead of accessing an HDF group directly
+
+    >>> hdf[group_name]
+    ...
+
+    you can wrap this with this class
+
+    >>> stub = HDFStub(hdf, group_name)
+
+    and then later perform this lookup with :method:`.load`
+
+    >>> stub.load() == hdf[group_name]
+    True
+
+    For simple datatypes there's not a big advantages to this, but :class:`.DataContainer` uses this to load its
+    contents lazily and ensure that nested containers are also lazily loaded.  This is done by customizing what happend
+    on :method:`.load` via :method:`.register`.  This class method adds a callback to the class that will be called
+    when the specified type name is found in the hdf group that is to be loaded.
+
+    >>> hdf['mytype/NAME']
+    MyType
+    >>> hdf['mytype/TYPE']
+    <class 'my.module.MyType'>
+    >>> HDFStub.register(MyType, lambda hdf, group: print(42) or hdf[group].to_object())
+    >>> my = HDFStub(hdf, 'mytype').load()
+    42
+    >>> my
+    MyType(...)
+
+    This is intended to allow classes that want to be lazily loaded in a certain way to customize what arguments they
+    pass `to_object()` (and therefore to their own initializers).
+    """
+
+    _load_functions = {}
+
+    def __init__(self, hdf, group_name):
+        """
+        Create new stub.
+
+        Args:
+            hdf (:class:`.ProjectHDFio`): hdf object to load from
+            group_name (str): node or group name to load from the hdf object
+        """
+        self._hdf = hdf
+        self._group_name = group_name
+
+    @classmethod
+    def register(cls, type, load):
```

Comment:
Finally, codacy complains that this overrides a built-in type...do you know what it's talking about? In pycharm I don't get any complaint, and I was expecting like when you try to use "id" or "dict" as a variable.

Reply:
The argument `type` is a builtin.  It is the base class of all classes and also a function to create new classes (iirc).  Since I'm not using this inside this short method, I think it's ok, but I can also come up with a different name.